### PR TITLE
7263 add diffusion loss

### LINKED ISSUE #7263: Add diffusion regularizer for image registration
**Is your feature request related to a problem? Please describe.**
Currently, for image registration task, MONAI only has [BendingEnergyLoss](https://docs.monai.io/en/stable/losses.html#bendingenergyloss) to enforce smooth deformations, which is based on the second partials of the displacement fields. A common regularizer used in unpaired brain MR images registration (e.g., [VoxelMorph](https://arxiv.org/pdf/1809.05231.pdf) and [TransMorph](https://doi.org/10.1016/j.media.2022.102615)) is the diffusion regularizer, which is based on the first partials of the displacement fields. It would be great if diffusion loss can be included as part of MONAI.

**Describe the solution you'd like**
Should be straight forward to implement using the existing `monai.losses.deform.spatial_gradient`.

**Describe alternatives you've considered**
One could alternatively customize a diffusion loss using `monai.losses.deform.spatial_gradient` in the training script. 

**Additional context**
Feel free to assign me to this issue. :)

### PR DESCRIPTION
**Pull request by kvttt**
Fixes #7263.

### Description

Add diffusion loss. I also made a [demo notebook](https://github.com/kvttt/deep-atlas/blob/main/diffusion_loss_scale_test.ipynb) to provide some explanations and analyses of diffusion loss. 

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by kvttt** [comment]
Hi @ebrahimebrahim, you worked on implementing `BendingEnergyLoss` earlier and provided an awesome demo notebook. Could you take a look at my demo notebook attached in this PR? It would be great to have you confirm that my analysis of diffusion loss make sense! Thank you!

**Comment by ebrahimebrahim** [comment]
Happy to take a look! I'll get back to you on this soon

**Comment by ebrahimebrahim** [comment]
I just had a look. I like your derivation of the scaling law for diffusion loss!
In your result
```math
L(\hat{{\bf u}}) = \frac{1}{|D|} \sum_{{\bf x} \in D} \left( \lambda_x^{-1}\lambda \odot  \frac{\partial}{\partial x} ({\bf u}({\bf x})) \right)^2 + \left( \lambda_y^{-1} \lambda \odot \frac{\partial}{\partial y} ({\bf u}({\bf x})) \right)^2 + \left( \lambda_z^{-1} \lambda \odot \frac{\partial}{\partial z} ({\bf u}({\bf x})) \right)^2
```
if you assume isotropic scaling $\lambda_x=\lambda_y=\lambda_z$ then there is cancellation and you've shown that $L(\hat{{\bf u}}) \approx L({{\bf u}}) $. That is, there's only a weird scale-dependent effect to worry about when you change your image resolution along different axes differently. The examples in the notebook always do isotropic changes of resolution. This explains why in your plot you see scale-invariance (the flat line) regardless of whether `normalize=True` or not:
image
To really see the benefit of your `normalize=True`, you'd want to change resolution differently on the x axis versus the y axis.

But yeah I think I agree with your derivation of the scaling law and I like your approach!

File: monai/losses/deform.py [hunk 1]
```diff
@@ -116,3 +116,74 @@ def forward(self, pred: torch.Tensor) -> torch.Tensor:
...
+                - ``"mean"``: the sum of the output will be divided by the number of elements in the output.
+                - ``"sum"``: the output will be summed.
+        """
+        super().__init__(reduction=LossReduction(reduction).value)
+        self.normalize = normalize
+
+    def forward(self, pred: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            pred: the shape should be BCH(WD)
```
[the remark this answers is not in the document]
Comment (ericspod) [reply monai/losses/deform.py:152]: And that the spatial dimension sizes must be greater than 2.

**Review by ericspod — APPROVED** [review]
Looks good, I had a few comments with minor changes.

**Comment by KumoLiu** [comment]
Would be great to add the demo to the tutorial repo and mention it in the docstring.

**Pull request merged**